File: VOID0000-api/server/vmd/storage.js
import { pool } from '../db.js';
import { ATTACH_BUCKET, minioClient } from '../minio.js';
import sentinel, { createSentinelKey } from '../sentinel/index.js';
import { resolveStoredAttachmentPolicy } from '../utils/attachmentContentPolicy.js';
import { transformVmdImage, VmdMediaError } from './imageVariants.js';
import { getVmdMetricsSnapshot, incrementVmdMetric } from './metrics.js';
import {
  createVmdVariantIdentity,
  vmdPersistentCache,
} from './persistentCache.js';
import { VmdWorkQueue } from './workQueue.js';

const DEFAULT_MAX_SOURCE_BYTES = 12 * 1024 * 1024;
const DEFAULT_MAX_CONCURRENT_TRANSFORMS = 2;
const DEFAULT_MAX_QUEUED_TRANSFORMS = 8;
const DEFAULT_QUEUE_WAIT_TIMEOUT_MS = 10_000;

function resolvePositiveInteger(value, fallback) {
  const parsed = Number(value);
  return Number.isSafeInteger(parsed) && parsed > 0 ? parsed : fallback;
}

function resolveNonNegativeInteger(value, fallback) {
  const parsed = Number(value);
  return Number.isSafeInteger(parsed) && parsed >= 0 ? parsed : fallback;
}

const MAX_SOURCE_BYTES = resolvePositiveInteger(
  process.env.VMD_MAX_SOURCE_BYTES,
  DEFAULT_MAX_SOURCE_BYTES,
);
const MAX_CONCURRENT_TRANSFORMS = resolvePositiveInteger(
  process.env.VMD_MAX_CONCURRENT_TRANSFORMS,
  DEFAULT_MAX_CONCURRENT_TRANSFORMS,
);
const MAX_QUEUED_TRANSFORMS = resolveNonNegativeInteger(
  process.env.VMD_MAX_QUEUED_TRANSFORMS,
  DEFAULT_MAX_QUEUED_TRANSFORMS,
);
const QUEUE_WAIT_TIMEOUT_MS = resolvePositiveInteger(
  process.env.VMD_QUEUE_WAIT_TIMEOUT_MS,
  DEFAULT_QUEUE_WAIT_TIMEOUT_MS,
);
const transformQueue = new VmdWorkQueue({
  maxConcurrent: MAX_CONCURRENT_TRANSFORMS,
  maxQueued: MAX_QUEUED_TRANSFORMS,
  waitTimeoutMs: QUEUE_WAIT_TIMEOUT_MS,
});
const CACHE_WARNING_INTERVAL_MS = 60_000;
const lastCacheWarningAt = new Map();

function warnCacheFailure(kind, error) {
  const now = Date.now();
  const lastWarningAt = lastCacheWarningAt.get(kind) || 0;
  if (now - lastWarningAt < CACHE_WARNING_INTERVAL_MS) return;

  lastCacheWarningAt.set(kind, now);
  console.warn(`[VMD_CACHE] ${kind}`, {
    error: error instanceof Error ? error.message : String(error || ''),
  });
}

function isObjectNotFoundError(error) {
  const code = String(error?.code || error?.name || '');
  return code === 'NoSuchKey' || code === 'NoSuchObject' || code === 'NotFound';
}

async function findAttachmentObject(attachmentId) {
  const result = await pool.query(
    `SELECT object_key
     FROM attachment_objects
     WHERE id = $1
       AND bucket = $2
     LIMIT 1`,
    [attachmentId, ATTACH_BUCKET],
  );
  return result.rows[0] || null;
}

async function readObjectBuffer(objectKey, expectedSize) {
  if (Number.isFinite(expectedSize) && expectedSize > MAX_SOURCE_BYTES) {
    throw new VmdMediaError('Attachment exceeds the VMD source limit', {
      code: 'VMD_SOURCE_TOO_LARGE',
      status: 413,
    });
  }

  let stream;
  try {
    stream = await minioClient.getObject(ATTACH_BUCKET, objectKey);
  } catch (error) {
    if (isObjectNotFoundError(error)) {
      throw new VmdMediaError('Attachment not found', {
        code: 'VMD_ATTACHMENT_NOT_FOUND',
        status: 404,
      });
    }
    throw error;
  }

  const chunks = [];
  let totalBytes = 0;
  try {
    for await (const chunk of stream) {
      const buffer = Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk);
      totalBytes += buffer.length;
      if (totalBytes > MAX_SOURCE_BYTES) {
        stream.destroy();
        throw new VmdMediaError('Attachment exceeds the VMD source limit', {
          code: 'VMD_SOURCE_TOO_LARGE',
          status: 413,
        });
      }
      chunks.push(buffer);
    }
  } catch (error) {
    if (error instanceof VmdMediaError) throw error;
    throw new VmdMediaError('Attachment storage read failed', {
      code: 'VMD_STORAGE_UNAVAILABLE',
      status: 502,
    });
  }

  return Buffer.concat(chunks, totalBytes);
}

async function renderStoredImage(attachmentId, variant) {
  const attachment = await findAttachmentObject(attachmentId);
  if (!attachment) {
    throw new VmdMediaError('Attachment not found', {
      code: 'VMD_ATTACHMENT_NOT_FOUND',
      status: 404,
    });
  }

  let objectStat;
  try {
    objectStat = await minioClient.statObject(ATTACH_BUCKET, attachment.object_key);
  } catch (error) {
    if (isObjectNotFoundError(error)) {
      throw new VmdMediaError('Attachment not found', {
        code: 'VMD_ATTACHMENT_NOT_FOUND',
        status: 404,
      });
    }
    throw new VmdMediaError('Attachment storage is unavailable', {
      code: 'VMD_STORAGE_UNAVAILABLE',
      status: 502,
    });
  }

  const sourcePolicy = resolveStoredAttachmentPolicy(objectStat, attachment.object_key);
  if (!sourcePolicy.inline) {
    throw new VmdMediaError('Attachment is not an approved sanitized image', {
      code: 'VMD_ATTACHMENT_NOT_SANITIZED',
      status: 415,
    });
  }

  const cacheIdentity = createVmdVariantIdentity({
    attachmentId,
    objectKey: attachment.object_key,
    objectStat,
    variant,
  });
  const cached = await vmdPersistentCache.read(cacheIdentity);

  if (cached.status === 'hit') {
    incrementVmdMetric('persistent_cache_hits');
    return cached.image;
  }

  incrementVmdMetric('persistent_cache_misses');
  if (cached.status === 'corrupt') {
    incrementVmdMetric('persistent_cache_corrupt');
  } else if (cached.status === 'unavailable') {
    incrementVmdMetric('persistent_cache_read_failures');
    warnCacheFailure('read unavailable; regenerating from source', cached.error);
  }

  let generated;
  try {
    generated = await transformQueue.run(async () => {
      const source = await readObjectBuffer(attachment.object_key, objectStat.size);
      const image = await transformVmdImage(source, variant);
      incrementVmdMetric('transforms_generated');
      return image;
    });
  } catch (error) {
    if (error?.code === 'VMD_AT_CAPACITY') {
      incrementVmdMetric('queue_full');
    } else if (error?.code === 'VMD_QUEUE_TIMEOUT') {
      incrementVmdMetric('queue_timeouts');
    }
    throw error;
  }

  try {
    generated.etag = await vmdPersistentCache.write(cacheIdentity, generated);
  } catch (error) {
    incrementVmdMetric('persistent_cache_write_failures');
    warnCacheFailure('write failed; serving generated response', error);
  }

  return generated;
}

export function renderStoredVmdImage(attachmentId, variant) {
  const flightKey = createSentinelKey('vmd.image-variant', attachmentId, variant);
  return sentinel.guard(
    flightKey,
    () => renderStoredImage(attachmentId, variant),
  );
}

export function getVmdStorageMetrics() {
  return {
    ...getVmdMetricsSnapshot(),
    work_queue: transformQueue.getSnapshot(),
  };
}

File: VOID0000-api/internal/vmd/storage.go
package vmd

import (
	"context"
	"errors"
	"fmt"
	"log/slog"
	"strings"
	"sync"
	"time"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"
	"github.com/minio/minio-go/v7"
	"github.com/minio/minio-go/v7/pkg/lifecycle"
)

type Storage struct {
	config      Config
	pool        *pgxpool.Pool
	minio       *minio.Client
	transformer *TransformClient
	queue       *WorkQueue[Image]
	flights     *FlightGroup[Image]
	metrics     Metrics
	logger      *slog.Logger
	warningMu   sync.Mutex
	warnings    map[string]time.Time
}

func NewStorage(config Config, pool *pgxpool.Pool, minioClient *minio.Client, logger *slog.Logger) *Storage {
	return &Storage{
		config:      config,
		pool:        pool,
		minio:       minioClient,
		transformer: NewTransformClient(config.TransformSocketPath, config.TransformTimeout, config.MaxSourceBytes, config.MaxVariantBytes),
		queue:       NewWorkQueue[Image](config.MaxConcurrentTransforms, config.MaxQueuedTransforms, config.QueueWaitTimeout, logger),
		flights:     NewFlightGroup[Image](config.MaxActiveFlights),
		logger:      logger,
		warnings:    make(map[string]time.Time),
	}
}

func (s *Storage) Metrics() map[string]any {
	return s.metrics.Snapshot(s.queue.Snapshot())
}

func (s *Storage) Shutdown(ctx context.Context) error {
	return s.queue.Shutdown(ctx)
}

func (s *Storage) warnCacheFailure(kind string, err error) {
	s.warningMu.Lock()
	now := time.Now()
	last := s.warnings[kind]
	if now.Sub(last) < time.Minute {
		s.warningMu.Unlock()
		return
	}
	s.warnings[kind] = now
	s.warningMu.Unlock()
	s.logger.Warn("VMD persistent cache degraded", "kind", kind, "error", err)
}

func (s *Storage) findAttachmentObject(ctx context.Context, attachmentID string) (string, error) {
	var objectKey string
	err := s.pool.QueryRow(
		ctx,
		`SELECT blob.object_key
		 FROM attachment_objects AS attachment
		 JOIN attachment_blobs AS blob
		   ON blob.id = attachment.blob_id
		 WHERE attachment.id = $1
		   AND blob.bucket = $2
		 LIMIT 1`,
		attachmentID,
		s.config.AttachmentBucket,
	).Scan(&objectKey)
	if errors.Is(err, pgx.ErrNoRows) {
		return "", mediaError(404, "VMD_ATTACHMENT_NOT_FOUND", "Attachment not found", nil)
	}
	if err != nil {
		return "", wrapInternal("find attachment object", err)
	}
	return objectKey, nil
}

func (s *Storage) statSource(ctx context.Context, objectKey string) (minio.ObjectInfo, error) {
	objectInfo, err := s.minio.StatObject(
		ctx,
		s.config.AttachmentBucket,
		objectKey,
		minio.StatObjectOptions{},
	)
	if err == nil {
		return objectInfo, nil
	}
	if isMissingObjectError(err) {
		return minio.ObjectInfo{}, mediaError(404, "VMD_ATTACHMENT_NOT_FOUND", "Attachment not found", err)
	}
	return minio.ObjectInfo{}, mediaError(502, "VMD_STORAGE_UNAVAILABLE", "Attachment storage is unavailable", err)
}

func (s *Storage) readSource(ctx context.Context, objectKey string, expectedSize int64) ([]byte, error) {
	if expectedSize > s.config.MaxSourceBytes {
		return nil, mediaError(413, "VMD_SOURCE_TOO_LARGE", "Attachment exceeds the VMD source limit", nil)
	}
	body, err := readBoundedObject(
		ctx,
		s.minio,
		s.config.AttachmentBucket,
		objectKey,
		expectedSize,
		s.config.MaxSourceBytes,
	)
	if err != nil {
		if isMissingObjectError(err) {
			return nil, mediaError(404, "VMD_ATTACHMENT_NOT_FOUND", "Attachment not found", err)
		}
		if strings.Contains(err.Error(), "size limit") || strings.Contains(err.Error(), "invalid size") {
			return nil, mediaError(413, "VMD_SOURCE_TOO_LARGE", "Attachment exceeds the VMD source limit", err)
		}
		return nil, mediaError(502, "VMD_STORAGE_UNAVAILABLE", "Attachment storage read failed", err)
	}
	return body, nil
}

func (s *Storage) renderStoredImage(ctx context.Context, attachmentID, variant string) (Image, error) {
	objectKey, err := s.findAttachmentObject(ctx, attachmentID)
	if err != nil {
		return Image{}, err
	}
	objectInfo, err := s.statSource(ctx, objectKey)
	if err != nil {
		return Image{}, err
	}
	if !trustedInlineImage(objectInfo) {
		return Image{}, mediaError(415, "VMD_ATTACHMENT_NOT_SANITIZED", "Attachment is not an approved sanitized image", nil)
	}

	identity, err := CreateCacheIdentity(attachmentID, objectKey, variant, objectInfo)
	if err != nil {
		return Image{}, wrapInternal("create VMD cache identity", err)
	}
	cached := readCachedImage(ctx, s.minio, s.config.CacheBucket, identity, s.config.MaxVariantBytes)
	switch cached.status {
	case "hit":
		s.metrics.persistentCacheHits.Add(1)
		return cached.image, nil
	case "corrupt":
		s.metrics.persistentCacheMisses.Add(1)
		s.metrics.persistentCacheCorrupt.Add(1)
	case "unavailable":
		s.metrics.persistentCacheMisses.Add(1)
		s.metrics.persistentCacheReadFailures.Add(1)
		s.warnCacheFailure("read unavailable; regenerating from source", cached.err)
	default:
		s.metrics.persistentCacheMisses.Add(1)
	}

	generated, err := s.queue.Run(ctx, func() (Image, error) {
		source, readErr := s.readSource(ctx, objectKey, objectInfo.Size)
		if readErr != nil {
			return Image{}, readErr
		}
		image, transformErr := s.transformer.Transform(ctx, source, variant)
		if transformErr != nil {
			return Image{}, transformErr
		}
		s.metrics.transformsGenerated.Add(1)
		return image, nil
	})
	if err != nil {
		var media *MediaError
		if errors.As(err, &media) {
			switch media.Code {
			case "VMD_AT_CAPACITY":
				s.metrics.queueFull.Add(1)
			case "VMD_QUEUE_TIMEOUT":
				s.metrics.queueTimeouts.Add(1)
			}
		}
		return Image{}, err
	}

	etag, err := writeCachedImage(
		ctx,
		s.minio,
		s.config.CacheBucket,
		identity,
		generated,
		s.config.MaxVariantBytes,
	)
	if err != nil {
		s.metrics.persistentCacheWriteFailure.Add(1)
		s.warnCacheFailure("write failed; serving generated response", err)
	} else {
		generated.ETag = etag
	}
	return generated, nil
}

func (s *Storage) Render(requestContext context.Context, attachmentID, variant string) (Image, error) {
	flightKey := fmt.Sprintf("%s:%s", strings.ToLower(attachmentID), variant)
	operationContext, cancel := context.WithTimeout(context.Background(), s.config.RenderTimeout)
	defer cancel()
	return s.flights.Do(requestContext, flightKey, func() (Image, error) {
		return s.renderStoredImage(operationContext, attachmentID, variant)
	})
}

func (s *Storage) InitializeCache(ctx context.Context) error {
	exists, err := s.minio.BucketExists(ctx, s.config.CacheBucket)
	if err != nil {
		return err
	}
	if !exists {
		if err := s.minio.MakeBucket(ctx, s.config.CacheBucket, minio.MakeBucketOptions{}); err != nil {
			return err
		}
	}
	if err := s.minio.SetBucketPolicy(ctx, s.config.CacheBucket, ""); err != nil {
		response := minio.ToErrorResponse(err)
		if response.Code != "NoSuchBucketPolicy" {
			return err
		}
	}
	lifecycleConfig := lifecycle.NewConfiguration()
	lifecycleConfig.Rules = []lifecycle.Rule{{
		ID:     "expire-vmd-derived-variants",
		Status: "Enabled",
		RuleFilter: lifecycle.Filter{
			Prefix: "variants/",
		},
		Expiration: lifecycle.Expiration{
			Days: lifecycle.ExpirationDays(s.config.CacheRetentionDays),
		},
	}}
	return s.minio.SetBucketLifecycle(ctx, s.config.CacheBucket, lifecycleConfig)
}

func (s *Storage) PingTransform(ctx context.Context) error {
	return s.transformer.Ping(ctx)
}
